tests/data_classes.py

```diff
@@ -0,0 +1,157 @@
+import itertools
+import os
+import json
+import subprocess
+
+import libzfs
+
+
+class ZFS(libzfs.ZFS):
+
+    def __init__(self):
+        super().__init(history=True, history_prefix='<iocage>')
+        pools = [
+            p for p in self.pools
+            if self.zfs_get_property(p.name, 'org.freebsd.ioc:active')
+        ]
+        self.pool = pools[0] if pools else None
+        assert self.pool is not None, 'Iocage pool not activated'
+
+    def _zfs_get_properties(self, identifier):
+        if '/' in identifier:
+            dataset = self.zfs.get_dataset(identifier)
+
+            return dataset.properties
+        else:
+            pool = self.zfs.get(identifier)
+
+            return pool.root_dataset.properties
+
+    def zfs_get_property(self, identifier, key):
+        try:
+            return self._zfs_get_properties(identifier)[key].value
+        except Exception:
+            return "-"
+
+    def get_dataset(self, dataset_name):
+        try:
+            return self.get_dataset(dataset_name)
+        except libzfs.ZFSException:
+            pass
+
+
+class Resource:
+    DEFAULT_JSON_FILE = 'config.json'
+
+    def __init__(self):
+        self.zfs = ZFS()
+
+
+# A simple class which can tell us the state of the jail
+class Jail(Resource):
+
+    def __init__(self, name):
+        super().__init__()
+        self.name = name
+
+    @property
+    def path(self):
+        # Jail can be either under `jails` or `templates` datasets
+        if self.zfs.get_dataset(
+                f'{self.zfs.pool.name}/iocage/jails/{self.name}'
+        ):
+            dataset = 'jails'
+        else:
+            dataset = 'templates'
+        return f'{self.zfs.pool.name}/iocage/{dataset}/{self.name}'
+
+    @property
+    def exists(self):
+        return self.zfs.get_dataset(self.path) is not None
+
+    @property
+    def template(self):
+        return 'iocage/templates/' in self.path
+
+    @property
+    def jid(self):
+        try:
+            return int(
+                subprocess.check_output(
+                    ['jls', '-j', f'ioc-{self.name.replace(".", "_")}'],
+                    stderr=subprocess.PIPE
+                ).decode('utf-8').split()[5]
+            )
+        except subprocess.CalledProcessError:
+            return None
+
+    @property
+    def running(self):
+        return self.jid is not None
+
+    @property
+    def config(self):
+        if os.path.exists(os.path.join(self.path, 'config.json')):
+            with open(os.path.join(self.path, 'config.json'), 'r') as f:
+                return json.loads(f.read())
```

Comment:
We should return `conf = iocage_lib.ioc_json.IOCJson(path).json_get_value('all')` Here as well, this currently will do no validation on invalid JSON.